### ISSUE TITLE
Support DaFang?

### ISSUE THREAD

**Issue by jejer** [samtap/fang-hacks#234]
Hi,

Will you support DaFang?
I want to use RTSP directly without MiHome.

Thanks!

**Comment by samtap** [comment]
I'm looking into it. I have one, and succeeded in getting serial console access. I've also intercepted a firmware update, but it's only a partial update not a complete firmware image.

Unfortunately it doesn't share any of the hardware components with XiaoFang. It is MIPS based instead of ARM, and has a different SoC. There don't appear to be many security provisions though (i.e. no signing like on Mijia1080p). It runs a telnetd on boot, which get killed immediately when Xiaomi software is started, so I don't have a simple 'jailbreak' method yet that doesn't require getting serial console access.

Any help or info is appreciated!

**Comment by ykhandler** [comment]
Try creating script in /etc/init.d
Ex: S07service
Contains: 
#!/bin/sh

busybox telnetd &

**Comment by samtap** [comment]
Sure, how do I create that without having serial console? ;-)

**Comment by ykhandler** [comment]
"I have one, and succeeded in getting serial console access", i thought you already have serial console access. Please share the firmware update file, maybe it has a clue.

**Comment by samtap** [comment]
I have serial console access but you can't expect regular users to take it apart and solder wires to the pcb.

We need to find something like the snx_autorun.sh trick in the older XiaoFang firmware. But the DaFang doesn't appear to use any scripts for mounting, same as on latest XiaoFang firmware they've integrated mounting into their executable instead of handling it using regular hotplug kernel facilities.

**Comment by ykhandler** [comment]
I see... may be someday i'll buy this camera and take a look of it...

**Comment by robinson4** [comment]
I just got one of these. I'm keen to get basic RTSP or even still image capabilities as well as GET requests to control PTZ functions.

I'm happy to help out with device testing etc. I have FTDI gear also.

**Comment by EliasKotlyar** [comment]
@samtap: Could you please provide a picture of the Pins on the Board? Will get my DaFang soon and want to try to investigate some details about the hardware.

**Comment by jimmyktp** [comment]
One of the reviews out there. https://www.youtube.com/watch?v=nDdnphZF6Cw

Seems like some function is not working. Currently, the latest firmware is 5.5.1.177, anyone tried that? I dare not to update the original firmware as I'm worried it might get shittier.

**Comment by EliasKotlyar** [comment]
Hello everyone,

Today i got mine. I havent managed to connect it yet, but have made a teardown. I collected all the informations in a new Repository.
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown

Feel free to contribute any information about the dafang.

Greatings,
Elias

**Comment by samtap** [comment]
Did you get a serial console? The pads are visible here https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/teardown/mainboard3.jpg on the side of the board you can easily add a header for gnd, rx, tx and route some wires through the bottom of the cam.

**Comment by EliasKotlyar** [comment]
Hello samtap,

Havent tried the serial connection yet. I need to get a FTDI-Adapter to try it out, and i will get it hopefully soon.

I just want to clarify about the serial connection:

Is the order of the Pins correct?
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/teardown/headers_sample.jpg

Could you please explain which bitrate and voltage you have used? I assume it is:
115200 Bits and 5V. 

Also i found out that there are some ports open which can be used to control this device. Will try MitM if i have time. Maybe we can control it using simple commands.

Greatings
Elias

**Comment by samtap** [comment]
Yes 115200, the level is 3.3V, I don't remember order of tx and rx pins but you can just switch them when it doesn't work ;)

**Comment by beikeland** [comment]
Would it be feasible to add a bluethooth serial module and find 3.3V on the board to avoid having to run wires. Anyone identified a 3.3V pad?

I'll gladly add serial headers - FTDI / Bluetooth to help test stuff when I get mine.

**Comment by EliasKotlyar** [comment]
Hello everyone,

I managed to get a FTDI with 3.3V and got a connection to the uboot. I have documented my way in my Repository.

Here is a bootlog of the device:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/mod_serial/booting.txt

Will try to dump the firmware soon/get a Telnet running. Any tips on that @samtap ?

@beikeland : It should work, there are a lot of 3.3V pads inside. Have a look on my solution - I just drilled a hole and routed some wires inside. 

Greatings
Elias

**Comment by Kadalia** [comment]
Hello,
I've just received a couple of Dafang ... unable to use then !
Read some forum comments, it seems that these cams are dedicated to china market :-(
definitively need a hack :-)
Go Elias Go !
I dont know how to help you .... so if I can do something for you, then just ask !
Thank you,
Philippe

**Comment by samtap** [comment]
If you disconnect flatcable to the sensor pcb, iCamera doesn't start since the sensor chip is missing, and you can see telnetd is running. It is killed by iCamera when it starts.

**Comment by nateinaus** [comment]
Hi, new to github so im not sure if sharing a file this way will work. I just received one of these and it turned up dead, I managed to track down the firmware from a guy on youtube. Not sure if its the full thing or not but it seemed to work for him. My camera doesn't even light up at all. The board is getting power and it warms up but nothing else. Hope this helps.
[demo_5.5.1.194.zip](https://github.com/samtap/fang-hacks/files/1404551/demo_5.5.1.194.zip)

**Comment by EliasKotlyar** [comment]
Hello everyone, 

I have several interesting news:

1. I have dumped the firmware and uploaded it here for further investigations.
2. There is some code in the bootloader (u-boot) which can make an whole update from the microsd. Unfortunately , i was not able to find out how to trigger it. Maybe someone can have a look? It starts with "jiabo":
jiabo_do_auto_update!!!!!!!!!!!!!!!!!!!!!!!!
jiabo_au_check_cksum_valid!!!!!!!!!!!!!!!!!!!!!!!!
jiabo_update_to_flash!!!!!!!!!!!!!!!!!!!!!!!!
jiabo_au_do_update!!!!!!!!!!!!!!!!!!!!!!!!
jiabo_idx=%d
jiabo_start=%x,jiabo_len=%x

I tried the binary from @nateinaus , but this code dont see it. 
@nateinaus : Can you please tell us the video of this guy on youtube?
@all: Could someone please ask Xiaomi if they can pass us the source of the u-boot? Its GPL, so they have to.  I know there is little chance that they will do it, but we still can try.

3. I have managed to crosscompile a dropbear server, but i cannot login. Its always says that it cannot find my user:
TRACE  (176) 29.525551: enter checkusername
TRACE  (176) 29.525586: leave checkusername: user 'root' doesn't exist

I dont see any error. Maybe someone can check?

Here is my build of the Dropbear:
https://github.com/EliasKotlyar/Dropbear-Dafang


Greating 
Elias

**Comment by Kadalia** [comment]
Hello,
I have "chat" with Xiaomi US team, they told me that Dafang was not yet available in US ... So they can't help me, and suggest me to mail to service.global@xiaomi.com.
Done ... wait for the source code of u-boot :-)
Very little chance .... but who knows !
Philippe

**Comment by nateinaus** [comment]
Yeah I thought it might just be an update to the firmware or something. I emailed xiaomi directly asking for the uboot but not sure if I'll have any luck. The guy on YouTube was a Spanish guy.

https://youtu.be/8ouH4_Pq4Ng

**Comment by ykhandler** [comment]
Nice,  sample make easier...

**Comment by Sniper-Wolf** [comment]
Hi there! 

Just offering my modest help : 
-> I can't help a lot with serial connections, but once we can get an easy telnet/ssh access, I can probably help with some scripts, dropping chinese server ip connections, replacing chinese voice files by english ones, and so on.

I messed around with the xiaomi mini router (i've since thrown it away, too much bugs). 

Here's a backup of the demo firmware from the video posted by nateinaus, just in case : 
https://drive.google.com/open?id=0B4WQ7BDk2rtZYy1Hb1dfaWE0TzA

5.5.1.194 suggests it's a newer firmware than the one often found in shipped webcams (usually 5.5.1.177 or lower)

**Comment by beikeland** [comment]
Just received my first camera; took some step-by-step pictures while opening it https://photos.app.goo.gl/Yjwh9TcIZiwLfXwL2

Is U2 / 25Q127CSIG  on the bottm an spi flash containing the firmware? Google didn't turn up much info on the part name, but could try a SOIC clip or desolder it and stick it in my CH341A and see if its able to pull some data.

**Comment by EliasKotlyar** [comment]
Hello everyone,

@Sniper-Wolf : Could you please have a look at my attemp to compile dropbear? 
https://github.com/EliasKotlyar/Dropbear-Dafang
I had the problem that could not login sucessfully. After that i found some patches(https://github.com/zcutlip/dropbear-hacks) and applied them. Now i can get past the login, but its segfaulting, and i cannot get reason for this error.

@beikeland I suppose it is the flash, but i am not sure. But you dont need to insert it into your programmer, because i already dumped the firmware here:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/tree/master/firmware/5.5.1.177

Here are some instructions for dumping it with serial connection:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/mod_getroot/getroot.md
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/mod_getroot/firmware-dump.md

**Comment by nateinaus** [comment]
Is there some way to reflash the whole firmware onto this camera using the sd card slot? I tried just copying all the files on and holding the setup button as I plugged it in, but nothing happened. Cheers

**Comment by EliasKotlyar** [comment]
Hello @nateinaus,

Unfortunately this is not possible at the moment. I have found some code in the bootloader(u-boot) which can do this, but i dont know how to make it work. If we would find a way to achieve this, it would be a huge step to a user-friendly flashing method. For now i can flash it only using serial-connection.

Greatings,
Elias

**Comment by beikeland** [comment]
@Sniper-Wolf if you'd be interested in remote access to the serial console I'd be happy to hook you with ssh access to a host with access to the serial console of the camera if it'll help us make progress. My email should be visible on my github profile for a while.

**Comment by EliasKotlyar** [comment]
I finally got this "firmware" from @nateinaus flashed on my Dafang. Many thanks @nateinaus!!

Here is a howto and my log:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/howto_firmwareflash.md

Warning: Dont try to do this unless you have a Serial Connection to rollback to the old one. I havent tested this firmware yet, i only know that its working for now.

I suppose that its not hard to modify it, because its a simple "uImage" Format.

**Comment by bilseto** [comment]
@EliasKotlyar for the demo firmware, have you tried the standard xiaofang root password?

**Comment by Sniper-Wolf** [comment]
Nice work Elias!

It's indeed an uImage file, i tried my luck on my Ubuntu VPS :
## First, you need to install u-boot-tools : 
_:~/dafang# apt install u-boot-tools_

Then download the firmware in a folder, and go into that folder.  (be sure not to rename it)
_:~/dafang# ls
demo_5.5.1.194.bin_

Then use this command : 
_:~/dafang# mkimage -l demo_5.5.1.194.bin
Image Name:   jz_fw
Created:      Sat Oct 14 05:03:40 2017
Image Type:   MIPS Linux Firmware (uncompressed)
Data Size:    11075584 Bytes = 10816.00 kB = 10.56 MB
Load Address: 00000000
Entry Point:  00000000
_

So we now know it's a linux firmware, from the 14 of October. It seems to be a single file image, as I do not have a "Contents" attribute. That's a kernel image, not a ramdisk image. 

We should be able to modify it but i haven't figured out how yet (I'm probably not the best for this task though).

Here's a link to help, but it's a bit different from our file : 
http://www.isysop.com/unpacking-and-repacking-u-boot-uimage-files/

**Comment by samtap** [comment]
The root password is the same as on XiaoFang: ismart12

**Comment by ykhandler** [comment]
@Sniper-Wolf, as i see the file. It contains kernel, root partition and jffs2 partition. So it is not only kernel image. And no kernel image filesize 3/4 portion compare with the flash space.

**Comment by Sniper-Wolf** [comment]
@ykhandler : Great, what do you use to see those partitions? Anyway to start modifying files in them, for example to try and keep the telnet alive?

**Comment by ykhandler** [comment]
Binwalk is the best i think... I got my dafang delivery delayed because custom checking. It should be arrived in 2-3 day from now. I'll see what i can do. Can't promise anything.

**Comment by EliasKotlyar** [comment]
I suppose that its in the same format as the normal firmware-dump:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/mod_getroot/firmware-dump.md

**Comment by kaputtzich** [comment]
One thing that's relatively easy is to extract two squashfs files out of the bin file und unpack them
`dd if=demo_5.5.1.194.bin bs=1 skip=2097216 count=3332360 of=file_1.squashfs`
`dd if=demo_5.5.1.194.bin bs=1 skip=5570624 count=564150 of=file_2.squashfs`
`unsquashfs file_1.squashfs`
`unsquashfs file_2.squashfs`
Now we have the filesystem...

**Comment by kaputtzich** [comment]
...or just use `binwalk -Me demo_5.5.1.194.bin` if too lazy :)

**Comment by EliasKotlyar** [comment]
Hello everyone,

I wroten a script to extract the firmware into parts:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/howto_unpackfirmware.md

Funfact:
The md5 of the kernel and the rootfs of my dumped firmware(5.5.1.177) matches with the binarys from the new "demo_5.5.1.194.bin". The appfs and the drivers different.

Greatings
Elias

**Comment by manjerico** [comment]
Hi all,

I have a couple of these Dafang Cameras and, although I can get them to work, sometimes I can't connect to them via 3G/4G.

Needless to say, that I would 1000000% prefer to connect directly to the cameras rather than via Xiaomi apps.

Unfortunately, I have no knowledge about programing/electronics to help, but I just wanted to send out my thank you and appreciation to those of you who are working on this!

If I can help in any other way, please let me know! I'll be hovering above this topic like a hawk. :p

Thanks!

**Comment by beikeland** [comment]
I was under the impression you had to have xiaomi router to access the cameras through the Mi app with the stock firmware?
(or is it just a matter of supporting uPnP?)

**Comment by manjerico** [comment]
I have a regular ISP-provided router. In my case, I can access the video feed and control the camera pan and tilt via Wi-Fi, when in the same network, every time (if you see in the MiHome App it calls it a Network Connection)

The problem is that when I try to connect via Mobile Data or via another network, sometimes it doesn't work. The Mi Home App just keeps on saying "Connecting (1/3)" and times out I guess. At the moment it's working fine, but later it might not. Some say that this is due to the fact this Remote function is not working well outside of China.

Nevertheless, and as I said, if a hack could be created, that would be golden! Accessing the camera directly instead of via the Mi Home app would be great.

If I can answer anything else, let me know. Thanks!

**Comment by ykhandler** [comment]
Ok, i think i successfully create the firmware. Will post it as soon as possible.

**Comment by EliasKotlyar** [comment]
Hey @ykhandler 

I have compiled the old firmware(5.5.1.177) . I need to try to flash it, but basically thats it:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/packer.py
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/demo_5.5.1.177.bin

Greatings
Elias

**Comment by ykhandler** [comment]
@EliasKotlyar , ok, You have to backup all partition before you tried it. I dont have the SDK, so i cannot compile the firmware, i just do modifying the file. 
I use the Backup data of my Camera to build it, because when i try using the nateinaus's firmware, the camera is loading so slow.
https://youtu.be/BF8UFTz5BoQ

**Comment by ykhandler** [comment]
I got a trick to run the telnet via the FIRMWARE.

**Comment by ykhandler** [comment]
i've update the FIMWARE with telnet enabled. I think that's it from me, i leave the project to you guys. Hope fully getting an RTSP, control the movement using telnet or else.

**Comment by EliasKotlyar** [comment]
Hello ykhandler,

Thanks for your firmware. As far as i have seen, you have modified the rootfs and integrated different soundfiles and a RC-Script for the Telnet:

```
#!/bin/sh

while [ 1 = 1 ]
do
sleep 1
result=`ifconfig wlan0 | grep UP`
if [ "$result" = "" ]
then
	cat /proc/uptime > /tmp/Handlerlog.txt
	echo "Running telnetd " >> /tmp/Handlerlog.txt
        echo "HANDLER RUNNING THE TELNETD now...."	
	/bin/busybox telnetD & 
	echo "telnetd should be Running, now killing check " >> /tmp/Handlerlog.txt
	break
	fi

done
```

**Comment by ykhandler** [comment]
@EliasKotlyar, yes i did, also i modify the string telnetd to telnetD in the busybox so that the iCamera cannot kill it.
Btw, the looping is wrong (it should be if [ ! "$result" = "" ]),and it must be not necessary to do the looping. 
I'm too lazy amd not know that i do it wrong until you show me the code.

Best regard

Handler

**Comment by kvanlier** [comment]
Hi @ykhandler,

Thanks for doing this.

Regards,

Krs

**Comment by manjerico** [comment]
Did you guys find something in the firmware that explain why sometimes you can't connect to the camera remotely and receive the video stream?

Thank you all once again!

**Comment by darioxxx89** [comment]
Hi guys, i'm very interested too for RTSP on this xiaomi dafang. This camera it's very cheap and very cool! The rotation works well, and if it is possibile to make rtsp fully working on this camera, i think it would be awesome. Any news about custom fw? I'm thinking to contact xiaomi and ask if they can add rtsp support maybe in future fw...it could be good idea?

**Comment by manjerico** [comment]
Any progress with this one? :)

**Comment by samtap** [comment]
Not until we get the SDK... And Ingenic have been ignoring my emails for weeks so I don't think we'll get it from them.

**Comment by EliasKotlyar** [comment]
Okey, i found something interessting that might help us:
On the chinese site, you actually can buy a development board with the T20 CPU:
http://www.ingenic.cn/?product/id/24/lm/3.html
https://item.taobao.com/item.htm?spm=a312a.7700824.w4004-14666727108.8.39637829p1DH9y&id=535966737504

Can someone who knows the language please translate the "conditions"? I suppose there is a NDA which should be signed before getting this boards. Also it might be helpful just to ask them on their special email-address: support_hf@ingenic.com

Greatings
Elias

**Comment by sogood007** [comment]
My understanding is that you can buy the development board but you need to sign the NDA for getting documentation.   However, it appears the board only ship within China. (from the taobao site)

**Comment by Kadalia** [comment]
Hello,
I don't know what exactly you are looking for .... but did you try the ftp site of ingenic ?
ftp://ftp.ingenic.com/DevSupport/Tools/Toolchain/ for instance.
I do not have find the SDK, but perhaps there are interesting things
Philippe

**Comment by PischeDev** [comment]
I already searched everywhere, nothing about the T20.

**Comment by ykhandler** [comment]
We're looking for SDK that might contains RTSPSVR code that we can compile and use...

**Comment by manjerico** [comment]
Are they obligated to share the SDK? (sorry, noob question)

Also, has anyone sent and email to the address found by @EliasKotlyar ?

**Comment by EliasKotlyar** [comment]
I might found another solution. Instead of compiling the rtsp, we might just take a precompiled from a similar device:

This IPCam has the same Ingenic T20 CPU as the Dafang:
https://www.aliexpress.com/item/S8816Y-WR-security-camera-Night-Vision-mini-CCTV-Camera-1080P-wi-fi-onvif-Wireless-IP-Camera/32748369585.html

Its being manufactured by the following company:
http://www.siepem.com/index.html

Can someone write an email in chinese to this company and kindly ask for a firmware and/or SDK? They might be little bit more cooperative. 

Greatings
Elias

**Comment by Kadalia** [comment]
https://github.com/dim08/Ingenic-T10_20

That ?

**Comment by PischeDev** [comment]
Well I guess it's the right one. Good job!
Spent 2 hours searching on Google and never thought about searching here.

**Comment by samtap** [comment]
I've taken a brief look at the SDK I think it's quite promising, but doesn't include a ready-to-use rtsp server. It does have some example code for the camera API so I'm thinking it should be possible to modify my XiaoFang rtsp server codebase to support both cams.

**Comment by manjerico** [comment]
These are great news!! :) :)
Once again, thank you all for your work!

**Comment by EliasKotlyar** [comment]
Hello everyone, 

I have managed to install this toolchain and to crosscompile the Dropbear-Server and a Live-Streaming using the live555(http://www.live555.com/mediaServer/#downloading) media server. Both of them are working but you need to integrate them into the firmware for later usage:

https://github.com/EliasKotlyar/Dropbear-Dafang

Greating
Elias

**Comment by MarcosStojan** [comment]
A new firmware was released today. 
Firmware version : 5.5.1.200

**Comment by samtap** [comment]
There's some kind of video server + windows client application in the SDK (/resource/tools/carrier), anybody try that?

**Comment by EliasKotlyar** [comment]
@samtap : I tried yesterday to compile the examples from the SDK. Unfortunately it haven`t worked out correctly yet. There was some code missing and i only can get a raw image for now. Getting a stream is failing with an error.

Here is my "toolchain" and my attemps:
https://github.com/EliasKotlyar/Dropbear-Dafang/tree/master/libimp-samples

**Comment by samtap** [comment]
The 'Carrier' server + client are prebuilt, no sources unfortunately, but would be nice to find out if they run on the camera, as a way to confirm the SDK is compatible with the current fw.

**Comment by jmtatsch** [comment]
Is it possible to downgrade the firmware?

**Comment by EliasKotlyar** [comment]
Sure. Just use an old version for flashing.

**Comment by MarcosStojan** [comment]
Anyone tried iSmartAlarm App from genuine manufacturers?
https://www.ismartalarm.com/devices/cameras/icamera-keep-pro/ISA00023.html

**Comment by kvanlier** [comment]
You cant switch to firmware from another cam as there is a hardware key (related to the manufacturer and product run) used in the encryption engine.

https://item.taobao.com/item.htm?spm=a1z10.1-c.w4004-14666727108.12.7bd1c079krQ569&id=538196665193

**Comment by manjerico** [comment]
Did you have any luck with the "Carrier" server and the SDK mentioned here?

*On a side note, do you guys have any idea of why the remote connection (mobile data/other wi-fi) sometimes does not work? Is it related to the connection to Chinese servers?

**Comment by iocatalinu** [comment]
Yeah,simply don't bother .Mine work 3 days perfect, 1 day slow, half day don't.This is the main reason i want direct rtsp stream,the  quality  of the image is awesome day and night,with/without IR.

**Comment by TheReallBeast** [comment]
> Yeah,simply don't bother .Mine work 3 days perfect, 1 day slow, half day don't.This is the main reason i want direct rtsp stream,the quality of the image is awesome day and night,with/without IR.

Same here (in the UK). When it works (not very often...) picture and functionality is great, shame it hardly ever connects. This is clearly not due to limited WIFI speed as I have fibre at 100Mb.

**Comment by EliasKotlyar** [comment]
Hello everyone,

Can someone please try to compile an openssh sftp server? It would help debugging a lot.  Also maybe can someone create a firmware especially for debugging and/or integrate all the scripts from the fang-hacks. I have written documentation how to achieve this and unfortunately don`t have time for getting everything together.


Greatings
Elias

**Comment by musicthebee** [comment]
Can you please post the documentation? I am willing to give it a shot. 

Thanks

**Comment by Silver360** [comment]
Hi, is posible with fang-hacks use night vision and motion detection? If yes, how it work?

**Comment by deennoo** [comment]
quote to follow this project

**Comment by blackgearit** [comment]
Hi all, following this discussion because I am dafang owner too and I don't like to send stream to xiaomi servers. But in the mi home app I've found that there's an direct wifi link to the mobile phone setup button.. Maybe the camera already expose a stream erver

**Comment by NishaYume** [comment]
Interested two dafang on the way and like others prefer direct rtsp conection instead pass through the xiaomi servers. Hope this can be possible.

**Comment by arn-aud** [comment]
@EliasKotlyar I have successfully built an OpenSSH server. Everything looks fine on the device, but I get errors when I try to open a session from my computer.
```
debug1: Connection established.
ssh_exchange_identification: read: Connection reset by peer
```
btw, isn't dropbear server enough for ssh/scp?

**Comment by EliasKotlyar** [comment]
Hello @arn-aud ,

Could you provide your solution to compile the OpenSSH? I tried it, but has no sucess yet. I was following the original Tutorial from the SDK. I am getting an error when trying to compile openssl against ulibc:

./libcrypto.so: warning: gethostbyname is obsolescent, use getnameinfo() instead.
./libcrypto.so: undefined reference to `getcontext'
./libcrypto.so: undefined reference to `setcontext'
./libcrypto.so: undefined reference to `makecontext'
collect2: error: ld returned 1 exit status
Makefile.shared:179: recipe for target 'link_app.gnu' failed
make[2]: *** [link_app.gnu] Error 1


Greatings
Elias

**Comment by arn-aud** [comment]
@EliasKotlyar https://gist.github.com/arn-aud/fef9e23b8f6e08416c8b18477d2226bb

**Comment by manjerico** [comment]
Just a heads up that the latest firmware (5.5.1.200, if not mistaken) allows to connect directly to the camera via Wi-Fi. That is, the camera creates an SSID and you can connect directly to it using the phone. Not sure if that helps in any way. :)

Obviously, I look forward to a more Xiaomi-independent solution. Thanks!

**Comment by sukoshi1507** [comment]
@manjerco: do you mean they added RTSP ?

**Comment by gysmo38** [comment]
@EliasKotlyar or @ykhandler I try to customize the firmware.
I manage to extract rootfs fomr ykhandler firmware.
I add live555server and dropbear files to /usr/bin and try to create new root-fs file.
But  mksquashfs generate a file with a diffrent size.
I also try to just extract with unsquashfs and do mksquashfs without modifcation in filesystem. I have the same issue.
Is there specific option to use? is it the right way to do it?

Thank for your help.

**Comment by EliasKotlyar** [comment]
Hello everyone,

@arn-aud 
>btw, isn't dropbear server enough for ssh/scp?

Dropbear dont support SFTP. It provides only a SSH-Connectio. You still need to build up an OpenSSH-Binary for having a SFTP. And SFTP will really make the debugging easier.

Many thanks for the Script. I will have a try with it.

@manjerico : I think that kinda every had this functionality - you just needed to press the button and then it was opening up a wifi-accesspoint.

@sukoshi1507 : No there is no RTSP Solution yet.

@gysmo38 : You can use some scripts to decompress/compress the firmware:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/packer.py

If your rootfs is bigger than the original one, it wont work. If its smaller, just pad everything with zeros until you get the desired size.

**Comment by arn-aud** [comment]
good news, we have a sftp server :) 
https://gist.github.com/arn-aud/fef9e23b8f6e08416c8b18477d2226bb

**Comment by gysmo38** [comment]
@EliasKotlyar Ok, my rootfs is smaller. Do you known how I pad with zeros?
There is an option to di it?
Thank a lot for your github, there is a lot of usefull informations.

**Comment by bjenmonk** [comment]
@gysmo38 
something like   dd if=/dev/zero bs=1 count=262144 >> largerfile.txt

**Comment by darioxxx89** [comment]
Great! is there any possibility to have custom fw to flash one day with rtsp or any other stream protocol in? :)

**Comment by gysmo38** [comment]
@balmurgan Thanks, it work. I manage to do my own firmware :)

**Comment by hdano1** [comment]
ready for testing :-)

**Comment by EliasKotlyar** [comment]
@gysmo38 : I have adjusted the packer.py script, now it will pad everything with zeros if the size is mismatched.
Have a try:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/packer.py

Here is a example how to use it:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/firmware/create_5.5.1.177.sh

**Comment by jbeck22** [comment]
what are the instructions for getting SSH working on this camera?

**Comment by effendiPL** [comment]
@EliasKotlyar ... any news?

**Comment by effendiPL** [comment]
@EliasKotlyar ... any news?

**Comment by ykhandler** [comment]
@gysmo38, have you succeed to create the rtspsrv?
@effendiPL, you're asking about what? 

for sftp, try this:
Flash his firmware : https://www.dropbox.com/s/oc5t8ycalibeozb/demo.bin?dl=0
Then run this binary using telnet:
https://www.dropbox.com/s/61xo0iiuzxm7s2q/sftp?dl=0

let me know is it working or not.

**Comment by gysmo38** [comment]
@ykhandler your live555media server work mays I do not manage to have picture with it. Does it work for you?
I try sftp but i get an error: "./sftp: can't load library 'libz.so.1'"

**Comment by ykhandler** [comment]
@gysmo38 , Have you flash the demo.bin i gave it with the sftp? Because i put the library inside the firmware.
I haven't tried your live555media server.

**Comment by EliasKotlyar** [comment]
Hello everyone,

I have great news:
1. I was able to compile the bftpd and now have a FTP-Connection to the Camera
2. With this connection i was able to compile the examples and get a h264 stream!

Now the bad part:
1. I was unable to make the sftp run. It is always failing when i try to connect. 
2. The stream is local only, so no RTSP for now
3. The stream is limited to 800x800 due using the second channel of the application.

As always , you can find my efforts in my repository:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Software

**Comment by darioxxx89** [comment]
Wow!!! great works!! hope in rtsp soon! ;) thank you for amazing job.

**Comment by ykhandler** [comment]
@EliasKotlyar, great job. Hope the rtspserver is done soon. I'm curious, how did you get the stream via the bftpd connection, can you explain a bit?

**Comment by EliasKotlyar** [comment]
Hello ykhandler,

The stream was written to a file, and then i have downloaded it. There is no streaming server yet, but i can get an h264 file in low resolution.

**Comment by ykhandler** [comment]
@EliasKotlyar, thanks for your explaination. Btw, have you figure it out, how the camera connect to wifi, without running the iCamera app? Maybe @samtap can also help us on this.

**Comment by doccaz** [comment]
Got two of these cameras here. One is working fine with the original 167 firmware, but the other one doesn't finish connecting with version 197. Does anyone have the 200 firmware available to download?
I intend to modify one of them with a WiFi serial as soon as I get some free time, and help with the rtsp.

**Comment by ykhandler** [comment]
Ok, i've able to connect to wifi without running iCamera, but by setting the ip manually,  is anybody successful compile the dhclient? Got my head dizzy cross compiling it.

**Comment by trisk** [comment]
My camera always claims to connect, and then says it failed a second later with both the original firmware (unclear which version it shipped with) and the patched 197. Wish there was an easier way to diagnose than setting up a serial console, maybe I should modify the image to log to sdcard.

**Comment by gysmo38** [comment]
@EliasKotlyar I try to crosscompile from yout gtihub but i have this error : "fatal error: selected multilib 'uclibc' not installed". Have you got an idea why? Thanks

**Comment by EliasKotlyar** [comment]
Hello everyone,

@ykhandler 
Take a look here: 
https://github.com/EliasKotlyar/Xiaomi-Dafang-Teardown/blob/master/mod_getroot/install_telnetandwifi.md
I havent compiled dhclient, but was still able to connect it to my home network.

@gysmo38 : Can you please post what exactly you are trying to compile? And how? For now i cannot help you, because there are to little information.

**Comment by gysmo38** [comment]
@EliasKotlyar  I just clone the project and execute compile_upload_libimp.sh

**Comment by EliasKotlyar** [comment]
@gysmo38 : Many thanks for the explanation. I have updated my repository. There were some files from the SDK  which were not checked in, because of the gitignore settings. Can you please try again? Now it should work

Greatings
Elias

**Comment by Lucas0511** [comment]
Greetings, anybody like samtap and me have both the Dafang and the Xiaofang? Is there a Mi Home version that supports both outside China? I can see the Xiaofang with 4.0.44 and the Dafang with the latest 5.0.10, but not the Xiaofang on 5.0.10 (region restriction), and the Dafang on the older 4.0.44 (not even starting). Of course once the bright minds here get us RTSP that problem could be solved otherwise. Thanks for your input.

**Comment by gysmo38** [comment]
@EliasKotlyar Thank it work :)

**Comment by jingle2008** [comment]
Mark, as I have some dafang cameras coming. Thanks for the hard work guys!

**Comment by myrmidondvb** [comment]
Full Busybox for MIPSEL
https://busybox.net/downloads/binaries/1.26.2-defconfig-multiarch/busybox-mipsel

work :)

sampel:  ./busybox-mipsel strings /system/bin/iCamera  > iCamara.txt

Greatings MyrmidonDVB

**Comment by ykhandler** [comment]
@EliasKotlyar. Thanks, it works. Dont need to compile dhcp anymore.

**Comment by ykhandler** [comment]
Hi @EliasKotlyar, i've compile your videostreamer, but why is it failed? Can you tell me what is wrong?
———————————————
sample_system_init start
failed to AddSensor
IMP_System_Init() failed
Play this stream using the URL "rtsp://192.168.100.166:7070/"
Beginning streaming...
Polling stream timeout

**Comment by EliasKotlyar** [comment]
Hello everyone,

@ykhandler : You have to insert the kernel module for the sensor before using that:
```
insmod /driver/sensor_jxf22.ko data_interface=2 pwdn_gpio=-1 reset_gpio=18 sensor_gpio_func=0
```

I got a RSTP Server which is running fine, but it has a lot of disadvantages for now:
1. Its not H264, its MJPEG
2. The Framerate ist quite bad(like 15 frames per second)
3. The Resolution is quite bad(only 640x480)

I will try to compile a H264 RTSP when i have free time.

Greatings
Elias

**Comment by manjerico** [comment]
Awesome work and dedication to all of you! :)

**Comment by inmenrinle** [comment]
Hi @EliasKotlyar 
Impressive work!!
I'm eagerly waiting to see your progress.
Just one question:
I'll be technically possible to control the pan and tilt functions in the future via RTSP or some web interface?
Thank you for your work and your time.

**Comment by darioxxx89** [comment]
@EliasKotlyar 

Thank you for your work!!! Awesome !!! :)

It would be amazing to have RTSP on this cheap camera...wait for new progress :)

**Comment by alcestneige** [comment]
@EliasKotlyar You are the man! Thank you for your dedication and hard work.

**Comment by EliasKotlyar** [comment]
Hello @inmenrinle 
It generally should be possible with some effors. I opened up a ticket about that topic:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/issues/1
Any help is welcome.

Greatings
Elias

**Comment by duelago** [comment]
Great work, guys. :) 
Looking forward to a RTSP-server with high res. I don't care about frame rate, but it would be nice to do a screen grab and ftp pictures up to a server every 15 minutes or so. I guess that this is not possible at the moment?

**Comment by macokk** [comment]
Hi! May I have questions about 'xiaomi dafang'?
It works, but I dont't understand a lot. I did not find a good manual. (a manual at all)
- in menu/camera settings/Auto record    - does it do anything? what? How can I see it? Where is it stored? 
- time lapse is recorded only to SD card? What happens, if wifi / power goes off?
Thanks

**Comment by sukoshi1507** [comment]
Dear @macokk; It Doesn't seem the purpose of @samtap 's github to explain how the Dafang works.

**Comment by macokk** [comment]
Then sorry for my question.

**Comment by EliasKotlyar** [comment]
@macokk:
This is actually possible. I have code which can capture a jpeg from the camera. Uploading to a FTP should not be a problem.

Also i was able to drive the horizontal axis with some compiled kernel driver.

**Comment by ykhandler** [comment]
@EliasKotlyar , wow, that is Fantastic, you can create your own application to control the Camera.

**Comment by duelago** [comment]
@EliasKotlyar Sounds great. Can you give me a hint how to capture a jpeg? :)

**Comment by EliasKotlyar** [comment]
Hello everyone,

@ykhandler : Not yet, it still needs some fine tuning. Also cannot drive it the vertical way(lacks of plugins)

@duelago : Sure, i have uploaded all my software into 
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/tree/master/firmware_mod/bin

Greatings
Elias

**Comment by liam-maps** [comment]
Hello @EliasKotlyar
I have a question for you.
Does your software replace the original firmware or just customize it?
I'm afraid that if I update the firmware version they can lock all of your patches.
Is it safe to update?
Is needed to have the latest version?

**Comment by EliasKotlyar** [comment]
Hello @liam-maps 

I havent released any custom firmware yet(work in progress for now). I only released some binarys which can be used to capture jpegs etc etc.

>Does your software replace the original firmware or just customize it?

My custom firmware will only allow to boot from the microsd instead of the original firmware. There will be no more customizations on the firmware level. Instead all the custom code will be on the microsd, and therefore can be updated really easy. 

>I'm afraid that if I update the firmware version they can lock all of your patches.

On my opinion they actually can release a new firmware which will lock down everything by introducing a new bootloader which will have some checks. However this is very unlikely, because this can become a really problem for them if some customers would "brick-flash" it by removing the power source etc etc. Also this product is not selling sooo good, that they will risk to make the customers angry.

> Is it safe to update?

For now on: Yes. I havent seen any firmware coming with a new bootloader.

>Is needed to have the latest version?

No. There is no such requirement.



Greating
Elias

**Comment by jmtatsch** [comment]
@EliasKotlyar how difficult would it be to get [motion](https://github.com/Motion-Project/motion) running directly on the dafang?

**Comment by EliasKotlyar** [comment]
Hey everyone,

I have managed to create a custom firmware. I need some testers, can someone test if its working? There are several things to fix, but i just want to get feedback of things which are working for now:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/master/hacks/install_cfw.md

@jmtatsch : Should be possible, but with a lot of efforts. You will have to compile a lot of librarys


Greatings
Elias

**Comment by hdano1** [comment]
I tried to update to your cfw, but not successfull. I am stuck  on the point nr. 6.
Dafang looks that it is updating, blue led lights up, but after about 1 minute device starts, but only empty folder time_lapse was created on the SD card, not the "firmware_mod" folder.
Any idea?

**Comment by hdano1** [comment]
@EliasKotlyar
I tried to update to your cfw, but not successfull. I am stuck on the point nr. 6.
Dafang looks that it is updating, blue led lights up, but after about 1 minute device starts, but only empty folder time_lapse was created on the SD card, not the "firmware_mod" folder.
Any idea?

**Comment by EliasKotlyar** [comment]
Hello @hdano1 ,

Thanks for testing. You need to copy the content of the firmware_mod folder to the root of the sdcard. I have updated the instruction, and i hope it will work then.

Greatings
Elias

**Comment by hdano1** [comment]
@EliasKotlyar
sorry, I thought that folder "firmware_mod" will be created by camera after first boot, and we need to copy it to root of the SD.
Now I got it :-)

**Comment by 2xdehelft** [comment]
I've come to the last step but my device won't connect to WiFi. Yellow light stays on (when starting without sd the light turns off..) fw installed successful, I had the newest version but not after the upgrade,

**Comment by EliasKotlyar** [comment]
@hdano1 : No the firmware-file has no major changes itself.. It will just allow to boot from microsd if some special files(run.sh) are provided. If you stick out the microsd or remove the run.sh, it will run as usual.

@2xdehelft : Do you have changed your ssid and the password in the config/wpa_supplicant file?

**Comment by 2xdehelft** [comment]
Yes I have, just ssid and key.. do I need to put them between ""? I have not changed any other setting like wpa. I know my network is set to support wpa and wpa2

**Comment by jasonbuechler** [comment]
Hi Elias -- I have two of these Dafangs that I got just for fun and I'm more than eager to test custom firmwares on them.  I just tried the following:

- brand new 64gb class u1 microsd card formatted as exFat
- downloaded cfw-1.0.bin, renamed it to demo.bin, and copied to the microsd card
- inserted the card into (unplugged) dafang, held down the setup button, plugged in the dafang, and released the setup button when the unit started spinning from the base
- the camera barked in annoyingly loud chinese for several minutes
- unplugged the camera, ejected the sdcard, and mounted the card on my windows pc
- like hdano1 said, an empty "time_lapse" folder had been created on the root level
- downloaded the entire project as a zip file, unzipped it, and copied the contents of the "firmware_mod" folder to the root of the sdcard
- modified e:\sample_config\wpa_supplicant.conf (using notepad++) to reflect my wifi's SSID and password (I used my 2.4ghz WPA/TKIP+AES network)
- re-inserted the sdcard into the dafang
- plugged the power cable back into the dafang
- unfortunately the LED is blinking yellow at about 4 blinks-per-second and does not appear as a client on my network

**Comment by 2xdehelft** [comment]
By the way the file is in sample_config?

**Comment by EliasKotlyar** [comment]
Hey, 
I have found several issues with the installing, and therefore have updated the installation instructions. Please delete the former CFW-1.0 and use 1.1. It has some indicator to check if the firmware was installed sucessfully.

@jasonbuechler : Unfortunately ExFat is not working for flashing. Consider using a FAT-Formatted Card.

@2xdehelft : You were right, there was some issue with the naming of the folders. I have corrected them, its now "config". My Idea was to support some "sample-configs" but for now on its not working as expected. I have removed this folder.

Greatings
Elias

**Comment by marvinroger** [comment]
@EliasKotlyar just wondering, why not `dd`-ing a "working" SD card into an `.img` that we could flash with a tool like [Etcher](https://etcher.io/)? It would avoid these kind of questions (exFAT, etc...).

**Comment by thoscut** [comment]
@EliasKotlyar great work, flashing and configuration went smooth, except I cloned the repo and the git client, working on a windows machine, modified the line endings from LF to CRLF. So downloading the zip archive of the master would be a helpful hint in the docs.

**Comment by jasonbuechler** [comment]
Success!!!
A new 2.4ghz device showed up on my network map, and browsing to its IP, I see "Webserver ist working!"

Here are my steps for future people:
- brand new 64gb class u1 microsd card
- used free version of Easeus Partition Master to format as FAT32
- downloaded cfw-1.1.bin, renamed it to demo.bin, and copied to the microsd card
- inserted the card into (unplugged) dafang, held down the setup button, plugged in the dafang, and released the setup button after about 10 seconds
- disconnected power from the camera as soon as the base started rotating (about 5mins)
- unplugged the camera, ejected the sdcard, and plugged the camera back in
- verified the LED shined blue for ~5 seconds (before, I think, turning yellow) before the base started moving
- unplugged the dafang
- mounted the sdcard on my windows pc
- downloaded the entire project as a zip file, unzipped it, and copied the contents of the "firmware_mod" folder to the root of the sdcard
- modified e:\config\wpa_supplicant.conf (using notepad++) to reflect my wifi's SSID and password (I used my 2.4ghz WPA/TKIP+AES network... I left the quotes in the text file, to those curious, @2xdehelft )
- re-inserted the sdcard into the dafang
- plugged the power cable back into the dafang
- camera appeared to boot up successfully and has a solid yellow LED
- camera showed up on my network, as seen from my router's web-admin console

@EliasKotlyar anything else you'd like me to test while I'm here? (also, the instructions still mention "sample_config" in the file-tree)
@tseel agreed on the hint to get the project zip. also recommend people use notepad++... cuz it's great.

**Comment by EliasKotlyar** [comment]
Hello everyone,

@marvinroger : Its a great idea - i will create a etcher image as soon this firmware become more stable. For now its testing only.

@tseel : Thanks for your feedback. I have put it into the configuration.

@jasonbuechler : You can try to check if these Urls are working:
http://DAFANGIP/cgi-bin/status
This should open up the fang-hacks-page

http://DAFANDIP/cgi-bin/currentpic
I hope you will get a picture in FULLHD here.

**Comment by jasonbuechler** [comment]
@EliasKotlyar both urls load (yay!) but the image appears to be corrupted/interrupted about halfway through each load ... attaching example here
![currentpic](https://user-images.githubusercontent.com/34217071/33529977-a12c3e84-d835-11e7-8832-80532bd05517.jpg)

**Comment by dbuxo** [comment]
thanks @EliasKotlyar, great work.  Both urls working, at first i got the same half image as @jasonbuechler, but then after several tries i saw that i can get a full image if the camera is not pointing all way down.

Three examples, with my second dafang with the same vertical position than the one recording.

![pointing_down](https://user-images.githubusercontent.com/708781/33530288-a58cd464-d87d-11e7-94c4-e5fad7105152.jpg)
![pointing_center](https://user-images.githubusercontent.com/708781/33530289-a9929116-d87d-11e7-8f48-5c20e088cbc0.jpg)
![pointing_up](https://user-images.githubusercontent.com/708781/33530290-abf7ca98-d87d-11e7-9617-540fcd607b7c.jpg)

**Comment by jasonbuechler** [comment]
Actually, I just made a bizarre discovery @dbuxo -- my camera HATES the window.  
Or to be more specific, something about the focus or brightness of pointing out my window "breaks" the image capture: if I point the camera directly out the window, then *nothing* will load.  If I point it at the frame of my window, but still get some out-the-window image, I will get a partial image load -- and it appears to be proportional (at least in the little testing I've done) yielding a more and more complete image, the more it is "avoiding" the window.

![window-frame1](https://user-images.githubusercontent.com/34217071/33530815-4db80b5e-d842-11e7-8179-aff9995b4418.jpg)
![window-frame2](https://user-images.githubusercontent.com/34217071/33530816-4dd5d580-d842-11e7-9c31-5d47e87d0b76.jpg)
![window-frame3](https://user-images.githubusercontent.com/34217071/33530817-4df30452-d842-11e7-982c-2533a2f96f62.jpg)
![window-frame4](https://user-images.githubusercontent.com/34217071/33530818-4e0f5436-d842-11e7-9dc0-39b30175838e.jpg)

**Comment by dbuxo** [comment]
yes @jasonbuechler , it looks like the error depends of what the camera sees (not the position of the axis).

Moving the camera around the room, i get mixed results, full picture, half picture and even no picture at all: 'the image cannot be displayed because it contains errors'

I don't know what triggers the error, but in these photos of my messy desk, a single bottle of coke on that exact spot brings the half picture error.  If i move the bottle the error is gone and i get a full picture.  

![pic_coke](https://user-images.githubusercontent.com/708781/33531503-49b6aa3c-d88e-11e7-9c17-a4a63088fac6.jpg)

![pic_no_coke](https://user-images.githubusercontent.com/708781/33531505-4c5ffd88-d88e-11e7-8906-e7c6db98cd54.jpg)

**Comment by dbuxo** [comment]
@EliasKotlyar on the boa.log file there is a error every time i get the half picture:

`Segmentation fault`
`192.168.1.40 - - [01/Jan/1970:00:26:32 +0000] "GET /cgi-bin/currentpic HTTP/1.1" 200 126976 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:58.0) Gecko/20100101 Firefox/58.0"`
`Bus error`
`192.168.1.40 - - [01/Jan/1970:00:26:42 +0000] "GET /cgi-bin/currentpic HTTP/1.1" 200 106496 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:58.0) Gecko/20100101 Firefox/58.0"`
`Bus error`

**Comment by deennoo** [comment]
Test is ok for me, using a 8go sd card.

I need to gitclone project under linux to get a demo.bin working
I don't need to wait about 5 minutes using cfw 1.1 to get base move (it takes aroud 120 sec maybe)
Connection is ok a wifi network running 2.4ghz & 5ghz under the same name.

http://DAFANGIP/cgi-bin/status : run as espered

http://DAFANDIP/cgi-bin/ : currentpic works mostly, have some half pics or no pics as other tester.

Thanx for your job.

**Comment by TheReallBeast** [comment]
Great work EliasKotlyar! Many thanks.
Does your latest mod include RTSP support please?

**Comment by liam-maps** [comment]
@EliasKotlyar you are the man!!!

**Comment by mcudogs** [comment]
I followed the instructions and both urls work as expected, great work, thanks.

**Comment by effendieffendi** [comment]
Great Work!! Any chance for RSTP?

**Comment by r0jx0** [comment]
It looks like mine doesn't connect to my wifi... how long does it take to connect to wifi?
I'll try again...

Anyway, thank you all for your work!

**Comment by EliasKotlyar** [comment]
Hello everyone,

I have just released some new Version of the binarys and added some features. Unfortunately , no h264 RTSP yet. But there are some other features to play around(like h264 recorder):
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/master/hacks/install_cfw.md

@dbuxo : I was able to reproduce this issue. There was some buffer which was to small. I have corrected this issue.

@everyone :  I have included a MJPEG RTSP server. I cannot proceed with the H264 one, and need help with that. Maybe someone have a look about that?
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/issues/3

@r0jx0 : It should connect instantly(like maybe 10 seconds)

**Comment by Karkas66** [comment]
I can confirm, the step by step instructions from @jasonbuechler works perfectly. The first dafang is connected to my wpa2 only 2.4ghz wifi. Statuspage is working, but most links are broken and you have to append ".cgi" manually. Static Image works, Mjpeg RTSP Server has to be tested yet, but first: Second dafang needs to be deployed.
@EliasKotlyar did you make a new firmware with your changes? It´s still v.1.1 in the download directory

**Comment by ril3y** [comment]
Off topic. Does anyone have an filesystem dump of the DaFang?  Mine are on
order.  I wanted to write support for it for my app'less configuration tool.

**Comment by EliasKotlyar** [comment]
@Karkas66 : I will provide only changes for the microsd-card. You dont need to flash anything other than FW1.1. 

About broken Links: Yeah, i have renamed everything to .cgi, to get it somehow sorted. I will have to fix broken links.  Besides, the other functionality is not so much developed. Maybe i disable it for now, until it gets fixxed.

@ril3y : Start here : 
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/tree/master/firmware_original

@everyone: Can someone test the LEDs, the mJpeg Stream and other stuff from new Version?

**Comment by duelago** [comment]
@EliasKotlyar Thank you! Your firmware is up and running. :) 
The camera stopped responding after the first time I took a snapshot. Rebooted and now it seems to work. Keep up the good work!

**Comment by jbeck22** [comment]
with the updated files do @jasonbuechler instructions still work if you are starting for the first time?  I have a camera at home and would love to get access working.

**Comment by Karkas66** [comment]
@EliasKotlyar LEDs are working, but you can not switch from blue to yellow, you have to switch blue off first. But mostly anoying is atm the "back" button, which does not work :P 
I Investigated in your RTSP Server but it does not get started, according to "netstat -a":
`[root@Ingenic-uc1_1:~]# netstat -a
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:21              0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:23              0.0.0.0:*               LISTEN
tcp        0    139 192.168.0.240:23        my-pc:51341           ESTABLISHED
Active UNIX domain sockets (servers and established)
Proto RefCnt Flags       Type       State         I-Node Path
unix  2      [ ]         DGRAM                       317 /var/run/wpa_supplicant/wlan0
[root@Ingenic-uc1_1:~]#`

Did I miss something?

**Comment by jbeck22** [comment]
@EliasKotlyar does your CFW allow me to set the proper time / date?

**Comment by Karkas66** [comment]
OK I found the hint, that you have to start the Streamer by entering "/system/sdcard/bin/mjpegStreamer 10" at the Commandline. But I can not Access the stream via VLC. I will test it with my [Motion](https://motion-project.github.io/) Server later

**Comment by 2xdehelft** [comment]
I've noticed the MAC adress changes when booting from SD. why is this? (thought it wasn't working cause i reserved an IP for the old MAC, but it changed...)

**Comment by r0jx0** [comment]
In the new version the led turns off after a while?

**Comment by deennoo** [comment]
@r0jx0 i notice this me too, a bit suprise at first time. maybe it have to stay yellow or blue or blink yellow and blue to confirm that wifi connection is ok.

@EliasKotlyar all is good for me

can you provied ssh password and login please ?

**Comment by thoscut** [comment]
@deennoo Use login credentials from here https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/master/hacks/getroot.md

**Comment by hdano1** [comment]
I hate the chinese voice commands of the camera.
Is it possible to replace it by update with next cfw?
Weeks ago, I tested cfw from youtuber ykhandler and it is cool.
@EliasKotlyar B,Y, IR  LEDs, motor left/right and audio are working, Get Picture  too.
Thank you !
Waiting for next updatessss :-)

**Comment by r0jx0** [comment]
@2xdehelft Thank you very much for your message I was searching the MAC address printed in the camera that is why I didn't find anything! Now I can connect.

@EliasKotlyar Everything is working fine though I don't know what the IR Led does, I don't see the difference in the pictures. Thank you very much for you work!

**Comment by jasonbuechler** [comment]
@deennoo login credentials for ssh/telnet/ftp = root/ismart12
@jbeck22 yes, since Elias just updated the contents of firmware_mod, my last set of instructions is still accurate
@EliasKotlyar  while SSH'ing to the camera, the connection seemed to drop every 1-3 minutes... it seemed to correspond to the camera also stopping responding to "currentpic" requests.

h264 dump seemed to work, functionally... however the output seems to be extremely low-res upon ftp'ing the file to my pc and playing VLC.  MediaInfo reports this specs: 
![image](https://user-images.githubusercontent.com/34217071/33630094-65c8f7a8-d9c3-11e7-9e68-d123f93daf20.png)
...320x240 pixels and it seems the image is highly zoomed-in, rather than resized

**Comment by Karkas66** [comment]
I checked if the new Version works with [MotionEyeOS](https://github.com/ccrisan/motioneyeos/wiki) on my Raspi but sadly neither the "Simple MJPEG Camera" nor the "Network Camera" on URL "rtsp://dafang:7070/" works. The last one actualy detects a "RTSP/Network Camera" but does not Show a picture and says "Unable to open video Device" instead.
It does not work with VLC too, so I think there is an actual problem with the mjpegStreamer binary

**Comment by EliasKotlyar** [comment]
Hello everyone,

@2xdehelft :
I have made a new issue about this. There is actually a way to achieve this.
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/issues/4

@r0jx0 : 
Yeah, i have fixxed this. Now the yellow LED will stay on.

@hdano1 : The chinese voice commands can be replaced like on ykhandler firmware. However it needs some efforts.

@r0jx0 : The IR-Led is for the nightvision mode. If turned on, it will provide light for the camera to see in the dark. You can test it using your mobile-phone-camera.  Just look at the dafang trought the mobile-phone camera and toggle the IR-Led.

Greatings
Elias

**Comment by samtap** [comment]
Are the voice samples the same as on XiaoFang? Quite likely they'd use the same voice samples for all cams that communicate and pair with mi home. I've made some new audio samples in english but I have a lot of trouble translating the chinese stuff with google translate, some audio samples don't really match what is happening in the app. So we could actually make some improvements here. And about half of the audio files never get trigger by using the app so no idea what to replace those with.
If anyone could provide a list, mapping all CN voice files to english voice files that would be very helpful.

A well-known trick to replace files on a read-only filesystem is bind mount, you don't need to replace the audio files you can just bind-mount them from sd on top of the existing files.

**Comment by EliasKotlyar** [comment]
Hello everyone,

i managed to create a RTSP-Server with h264. For now its for debugging only, cause its being really slow. But i will try to improve it soon. 
Also that issue with the MAC-Address been fixxed. Many thanks @dbuxo !

Greating
Elias

**Comment by jbeck22** [comment]
got the CFW installed this morning.  

So far the following items work OK:

- LED lights
- get current image **gets the current pic right away, but will not allow me to refresh later
- rotate left and right
- IR lights on / off


Things that don't work for me:

- Setting the name or timezone -- gives error: **_Unsupported command 'settz'_**
- Getting the current picture stops working after just a few secs and requires a reboot
RTSP gives me this error:
_Play this stream using the URL "rtsp://<IP Address>:7070/"
Beginning streaming...
Polling stream timeout
Segmentation fault_ 


**Feature Request**:

- Fix above issues
- URL video access -- something like http://IP_address/cgi-bin/video.cgi
- auto enable the IR controls based on ambient light
- vertical motor control

**Comment by sogood007** [comment]
@samtap If someone can provide me the pointer to the audio files, I can translate them to English "text". 
 Other volunteers can use that to create the audio other language files.

**Comment by duelago** [comment]
I want to give my camera a static IP. Can someone please tell me how to fix this this? I'm running the excellent @EliasKotlyar firmware :)

**Comment by poof86** [comment]
Giving a device a static IP is something you can do in your router settings. Check the DHCP server settings and usually you can find your device there and make the ip static :)

**Comment by sharkoz** [comment]
@sogood007 I found some .wav files in /usr/share : 
usr/share/Nokia.wav
usr/share/baojing.wav
usr/share/notify/CN/code_ok.wav
usr/share/notify/CN/code_wrong.wav
usr/share/notify/CN/connect_wifi_ok.wav
usr/share/notify/CN/connecting.wav
usr/share/notify/CN/copy.wav
usr/share/notify/CN/dang.wav
usr/share/notify/CN/get_msg.wav
usr/share/notify/CN/init_ok.wav
usr/share/notify/CN/speaker.wav
usr/share/notify/CN/update_system.wav
usr/share/notify/CN/upgrade_success.wav
usr/share/notify/binbin.wav
usr/share/notify/dang.wav

**Comment by duelago** [comment]
@poof86 Nope. My 4G router don't support this. Really stupid.

**Comment by jbeck22** [comment]
how do you get access to FTP?

**Comment by EliasKotlyar** [comment]
@jbeck22 : same login as in ssh

**Comment by jbeck22** [comment]
@EliasKotlyar thanks --- got it working.

can you help me get the mjpeg stream working?  Still getting the segmentation fault when I try to start it.

**Comment by EliasKotlyar** [comment]
@jbeck22 : Update to the last version. It includes a rtso h264 server, with a start button in the status-page

**Comment by sharkoz** [comment]
Up and down controls for the motor don't work for me.

Also, once the hack is installed, can I re-do the setup to access the cam from the Xiaomi app ?

By the way : Thanks for your amazing work 👍

**Comment by EliasKotlyar** [comment]
@sharkoz : Many thanks for the PR. I merged it already. About the up/down: Can you describe whats wrong?

@everyone: Can we close this issue down? The fanghacks have already been ported . I have opened up another issue for general discussions in my repo:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/issues/7

**Comment by jbeck22** [comment]
@EliasKotlyar updated to the latest.

Up Down Left and Right now work!  Thanks for that!

RTSP isn't working -- when I press the button nothing happens.

When I try to start RTSP from the command line this is what I get:
_**[root@Ingenic-uc1_1:bin]# ./mJpegStreamer 10
Play this stream using the URL "rtsp://10.0.0.199:7070/"
Beginning streaming...
Polling stream timeout
Segmentation fault**_

**Comment by nwdcben** [comment]
@2xdehelft I'm also having issues finding my Dafang once it reconnects to wifi after editing the WPA config file.  You indicated it had a new MAC address, is it close to the original?  I have too many devices on my network to identify it, if it is using a new MAC.  Do you mind posting your original and new mac; you can XX out any characters that are the same.  Thanks!

**Comment by nwdcben** [comment]
Disregard, I just figured it out.  It does have the SAME MAC address before and after.  For everyone struggling, don't use a router that has the same SSID for both 5GHz and 2.4GHz; that was my issue.

**Comment by Karkas66** [comment]
There is an Error in the status.cgi where you start the RTSP Server.

**Comment by Justalurker1213** [comment]
Mine doesn't seem to want to flash the firmware. I rename to demo.bin and put on root of FAT formatted SD card, pull power from camera, insert SD, hold setup, plug in power but it never blinks blue. It blinks yellow for about 15 seconds and then the base starts moving and i get the chinese prompts. What might I be doing wrong?

**Comment by markus0m** [comment]
Same happened to me at first. Try a different card and make sure to properly format it before.

**Comment by marvinroger** [comment]
It would be useful to maintain a Wiki page/MD document of known working SD
cards.

**Comment by lolouk44** [comment]
Thanks @EliasKotlyar !
I know the RTSP is still a work in progress, so for now I'm using the GetImage URL with motion.
It's a bit slow as I only get ~1 frame every 2 sec, but it works.
Looking forward to get real RTSP now :)

**Comment by htavasoglu** [comment]
First of all, thank you so much @EliasKotlyar 
I followed your directions and installed the custom firmware successfully.
At the moment the only problem is I can't get rtsp feed using ip cam viewer software. 
When I use VLC with rtsp://blah blah..../testStream it perfectly gets the stream. It is a little low resolution but usable. When I use the same URL with Android IP cam viewer or Windows IP cam viewer software they can't show the stream. 
I don't know how I can give more detail....
Is there a known difference between VLC and IP cam viewer apps on showing the rtsp feed?

**Comment by lolouk44** [comment]
Wondering whether we'll get some RTSP For Xmas? That would be a great gift to end the year :)

**Comment by rodrigoscoelho** [comment]
@EliasKotlyar , just to know. Do you have any idea if will be possible to have full RTSP sometime in future? 
What is your opinion since you know the insides better than us.

**Comment by EliasKotlyar** [comment]
@rodrigoscoelho : I already have a working RTSP. Its just in the dafang-hacks repository:
https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks

I dont know what exactly you are expecting/missing in the actual version. It supports both H264 and MJPEG. Moreover, its based on 
https://github.com/mpromonet/v4l2rtspserver
and supports all kind of parameters.

**Comment by rodrigoscoelho** [comment]
Ohh Great. I didn't see that.  I was watching this forum and the latest info was about RTSP only for JPEG. Great to know. Thank you so much.

**Comment by EliasKotlyar** [comment]
Yeah, i think its time to close this issue - we have a full support for the Dafang.

@samtap Could you please close this issue? Many Thanks!

**Comment by rodrigoscoelho** [comment]
on the page, there is:

"RTSP with mJPEG (low quality):
/system/sdcard/bin/mjpegStreamer 10
Local h264 recording:
/system/sdcard/bin/h264Snap > /system/sdcard/video.h264"

Only local h264 recording?

**Comment by rodrigoscoelho** [comment]
RTSP only with mJPEG?

**Comment by EliasKotlyar** [comment]
@rodrigoscoelho : I have corrected this. It was not updated during the development.

**Comment by rodrigoscoelho** [comment]
perfect. Thanks Elias

**Comment by bmwm69** [comment]
Hello, 

that RTSP works with HA ? Home Assistant ? Tks

**Comment by hdano1** [comment]
Hi all,

I updated latest cfw, I have all files on SD card, I can connect to the web interface of the camera.
But I would like to use camera with some app in android to watch the video + move with it.
Any guide, how to get this? Which app can support this?
Possible to connect to the camera in AP mode? Without router ?
Thx for reply.

**Comment by lolouk44** [comment]
@bmwm69 I use [motion](https://github.com/Motion-Project/motion) for this as RTSP in HA is not the most stable

**Comment by EliasKotlyar** [comment]
@hdano1 : You can use any RTSP capable client on Android to watch the stream. The PTZ functionality (moving around the camera) wont work.

**Comment by WOnken** [comment]
As for me, everything works but not the webserver. I can connect via SSH and so on but the wenserver doesn't work (on port 80)...

**Comment by effendieffendi** [comment]
HI, how to check the firmware version on camera?

**Comment by WOnken** [comment]
Before using the custom firmware I upgraded to the latest update > 200. But doesn't the custom firmware replace the existing with Version 5.5.1.200 anyways?

**Comment by jmtatsch** [comment]
@effendieffendi check http://dafanghacks/cgi-bin/status.cgi for the version the custom fw is based upon
@WOnken The latest custom fw is also based on 5.5.1.200, so no need to update with xiaomi app first.

**Comment by olskar** [comment]
What are the latest instructions for installing latest custom fw on this device?

**Comment by jmtatsch** [comment]
@olskar have a look at the [repo supporting the dafang camera](https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/master/hacks/install_cfw.md)

**Comment by olskar** [comment]
thanks @jmtatsch ! got it to work :)

**Comment by Dgak10** [comment]
Hi
I must be a bit stupid for not getting to work with https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/master/hacks/install_cfw.md
In general information it says that we have to put a run.sh file in the sd card
In installation of cfw that There should be no other files on the microSD
I dont understand why a i have to installa a new fw in Installation of the new Firmware if i already have the CFW ?

Yes i'm lost ;)

thanks for your help

**Comment by ykaminoh** [comment]
@Dgak10 
It's 2 separate steps, 1st step to install the custom bootloader (demo.bin), then 2nd step to run the custom firmware (run.sh).

**Comment by ron-ald52** [comment]
Hi

Is anyone here working on the tracking capability the original firmware has? I'd like to have it setup such that:

if movement is detected capture image either to SDcard or emailed, and/or make accessible from the UI

Thanks

**Comment by fincanbill** [comment]
Hi Evry mode when motion comes on my ios apple iphone 6 motion is waiting short time then %100 not showing how can I solve this problem ?

![185853d7pk4ljfc7jxjz77 jpeg thumb](https://user-images.githubusercontent.com/36035423/35732258-d5b87f5e-0829-11e8-9f75-965ba7b1487c.jpg)
![185949mfgttg9iw9mee0mf jpeg thumb](https://user-images.githubusercontent.com/36035423/35732259-d5d85806-0829-11e8-9a34-c5d6b5e473f7.jpg)

**Comment by zgadgeter** [comment]
I'm stil not able to flash my device. I have tried 3 different SD cards, so do not think it is a problem with the SD card.
What I do is as follows:
I format my SD card with FAT32
I copy the demo.bin onto the car. There are no other files on the card.
I put the card into the camera.
I hold the setup button on the camera
I then plug in the camera to power.
I have tried holding the setup button for a long time, and also have let it go after powering the camera up, same result.
The camer powers up, and with about 10 seconds begins to turn it's base as though it's starting up normally.
So, does anyone have any further suggestions?

thanks.

**Comment by effendieffendi** [comment]
Hi, how to remove name, data&time signature from rtsp stream?

**Comment by shinebar1001** [comment]
Please ask this questions in the Issues of Dafang-Hacks, they will be answered faster.
At the moment you have to edit run.sh on the sdcard 
Change

`/system/sdcard/bin/busybox nohup /system/sdcard/bin/v4l2rtspserver-master -D "DafangHacks %I:%M:%S %d.%m.%Y" -d UP &>/dev/null &`

to
`/system/sdcard/bin/busybox nohup /system/sdcard/bin/v4l2rtspserver-master &>/dev/null &`

**Comment by kevincaradant** [comment]
I@zgadgeter, Hi! , I know it's not the good location for that but did you find the answer ? I have exactly the same issue ! I'm currently running with the lastest firmware .287. Thank you

Edit: problem solved using format under Linux a low level in fat32 :)
